x/staking/migrations/v045/store.go

```diff
@@ -0,0 +1,56 @@
+package v045
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	v040staking "github.com/cosmos/cosmos-sdk/x/staking/migrations/v040"
+)
+
+// MigrateStore performs in-place store migrations from v0.43/v0.44 to v0.45.
+// The migration includes:
+//
+// - Removing delegations that have a zero share or token amount.
+func MigrateStore(ctx sdk.Context, storeKey sdk.StoreKey, cdc codec.BinaryCodec) error {
+	store := ctx.KVStore(storeKey)
+
+	return purgeDelegations(store, cdc)
+}
+
+func purgeDelegations(store sdk.KVStore, cdc codec.BinaryCodec) error {
+	prefixDelStore := prefix.NewStore(store, v040staking.DelegationKey)
+
+	delStoreIter := prefixDelStore.Iterator(nil, nil)
+	defer delStoreIter.Close()
+
+	valCache := make(map[string]v040staking.Validator)
+
+	for ; delStoreIter.Valid(); delStoreIter.Next() {
+		var delegation v040staking.Delegation
+		if err := cdc.Unmarshal(delStoreIter.Value(), &delegation); err != nil {
+			return err
+		}
+
+		validator, ok := valCache[delegation.ValidatorAddress]
+		if !ok {
+			valAddr, err := sdk.ValAddressFromBech32(delegation.ValidatorAddress)
+			if err != nil {
+				return err
+			}
+
+			if err := cdc.Unmarshal(store.Get(GetValidatorKey(valAddr)), &validator); err != nil {
+				return err
+			}
+
+			valCache[delegation.ValidatorAddress] = validator
+		}
+
+		// TODO: On-chain, we call BeforeDelegationRemoved prior to removing the
```

Comment:
@ValarDragon @amaurym I'm not sure if we need to mimic the `BeforeDelegationRemoved` hooks here? What this does on-chain is that certain distribution calls are made. My gut tells me we can avoid these here. Just wanted to confirm.

Reply:
I don't think hooks should matter on upgrades, but let me double check the hook usage